### ISSUE TITLE
Widening after inferring null or undefined for a type parameter should be an implicit any

### ISSUE THREAD

**Issue by mhegazy** [microsoft/TypeScript#531]
Courtesy of @JsonFreeman 

```
function f<T>(t: T): T { return null }
f(null);
```

When compiling with -noImplicitAny, we don't get an error here, even though T is implicitly inferred to be any when we call f

**Comment by DanielRosenwasser** [comment]
Bumping this since we ran into a situation with a leaked `any` in our own compiler. Minimal example:

```TypeScript
function forEach<T, U>(array: T[], callback: (element: T, index: number) => U): U {
    if (array) {
        for (let i = 0, len = array.length; i < len; i++) {
            let result = callback(array[i], i);
            if (result) {
                return result;
            }
        }
    }
    return undefined;
}


forEach([1, 2, 3, 4], () => { 
    // do stuff
    if (blah) {
        // bam! accidental any
        return undefined;
    }
});
```

**Comment by mattmccutchen** [comment]
FWIW, I just missed an implicit-any error in [my code](https://bitbucket.org/objsheets/objsheets/src/92de3a2de7b9af7caf53943ccce518dff297b184/src/client/sheet.ts?at=master&fileviewer=file-view-default#sheet.ts-228) due to this issue:
```ts
// Initialize a "height x 0" array of arrays.
let grid = _.range(0, height).map((i) => []);
// Append to rows...
```
The type of the empty array was inferred as `undefined[]` and then widened to `any[]` without error.  (Hopefully adding some keywords will help the next person who hits this with an empty array but isn't familiar with widening to find this issue.)

**Comment by JsonFreeman** [comment]
As I understand it, there has been a big shift to better integrate `null` and `undefined` into the type system, though I'm not up on the details. In 2014, when this issue was last looked at, these types were sort of the dark corners of the type system. I think it is worth asking whether this widening step still needs to occur given the latest semantics for null and undefined types.

**Comment by UselessPickles** [comment]
I just encountered a much simpler situation where I expected "noImplicitAny" error:

```ts
// Type of 'something' is 'any'! No error :(
let something = null;
```

Is this the same issue, or should I create a new one?

**Comment by mattmccutchen** [comment]
@UselessPickles: That `any` doesn't indicate a type safety problem, since TypeScript's control flow analysis will compute a more specific type for the variable each time it is used.  If you like, you can file an issue (after checking for an existing one) to have the language service show something more informative than just `any`.

**Comment by Andarist** [comment]
I don't think this is a problem anymore. The code is fine even on the oldest TS version available in the playground, see [here](https://www.typescriptlang.org/play?ts=3.3.3#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIB4AVAPgAoMAueIgSmqIChEzVkIJaBueIA). It might be worth to close this issue.

**Comment by RyanCavanaugh** [comment]
Making a call here that no one should have `noImplicitAny` on and `strictNullChecks` off (or vice versa). There are too many behaviors that are nonsensical when these flags don't align. It won't be a banned configuration per se but it's not a configuration that you can reason about.
